core/src/main/java/org/apache/iceberg/data/GenericRecord.java

```diff
@@ -115,15 +120,27 @@ public Object get(int pos) {
   }
 
   @Override
+  @SuppressWarnings("unchecked")
   public <T> T get(int pos, Class<T> javaClass) {
     Object value = get(pos);
     if (value == null || javaClass.isInstance(value)) {
       return javaClass.cast(value);
+    }
+
+    if (value instanceof LocalDateTime) {
+      return (T) (Long) timestampToMicros((LocalDateTime) value);
+    } else if (value instanceof OffsetDateTime) {
+      return (T) (Long) timestampToMicros(((OffsetDateTime) value).toLocalDateTime());
     } else {
       throw new IllegalStateException("Not an instance of " + javaClass.getName() + ": " + value);
     }
   }
 
+  private long timestampToMicros(LocalDateTime value) {
```

Comment:
It's best to avoid adding new local conversions. We prefer to use the well-tested ones in `DateTimeUtil`.